File: 15.3Sum/main.cpp
#include<iostream>
#include<vector>

using std::vector, std::cout, std::endl, std::string, std::begin, std::end;


class Solution {
 public:
  vector<vector<int>> threeSum(vector<int> nums) {
    vector<vector<int>> ans;
    sort(nums.begin(), nums.end());

    for (int i = 0; i < nums.size(); i++) {
      if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
        int low = i + 1;
        int high = nums.size() - 1;
        int temp = 0 - nums[i];
        while (low < high)
          if (nums[low] + nums[high] == temp) {
          vector<int> st;
          st.push_back(nums[i]);
          st.push_back(nums[low]);
          st.push_back(nums[high]);
          ans.push_back(st);
          while (low < high && nums[low] == nums[low + 1])
            low++;
          while (low < high && nums[high] == nums[high - 1])
            high--;
          low++;
          high--;
          } else if (nums[low] + nums[high] > temp) {
            high--;
          } else {
            low++;
          }
      }
    }
    return ans;
  }
};

int main() {
  for (auto i : Solution().threeSum({1, 2, -3, 3, 0})) {
    for (auto j : i)
      cout << j << ' ';
    cout << '\n';
  }
  return 0;
}

File: 15.3Sum/main.py
class Solution:
    def twoSum(self, nums: List[int], idx: int, res: List[List[int]]):
        seen = set()
        fast = idx + 1
        while fast < len(nums):
            complement = -nums[idx] - nums[fast]
            if complement in seen:
                res.append([nums[idx], nums[fast], complement])
                while fast + 1 < len(nums) and nums[fast] == nums[fast + 1]:
                    fast += 1
            seen.add(nums[fast])
            fast += 1
        
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for left in range(len(nums)):
            if nums[left] > 0:
                break
            if left == 0 or nums[left - 1] != nums[left]:
                self.twoSum(nums, left, res)
        return res
